File: python/bench_range.py
#!/usr/bin/env python3
# Copyright (C) Advanced Micro Devices, Inc.
# SPDX-License-Identifier: MIT

"""Python range benchmark: what an idle range costs when no tool is attached.

The C++ benchmark (benchmarks/bench_range.cpp) makes one point: eager argument
evaluation is the entire cost, and a lambda removes it. Python makes two, and
the second is the reason this file exists.

  1. Deferred arguments matter here too, for the same reason -- an f-string
     passed as an argument is built before the range is ever entered.

  2. Unlike C++, the scope machinery itself is not free. A guarded pair of raw
     calls is the floor; the decorator adds a *args/**kw wrapper on top of it,
     and the context manager adds an object allocation plus two method calls.
     No amount of deferring removes that. It is the price of the syntax.

Every measurement below is taken with NO tool attached, which is the case that
matters: it is what an instrumented application pays in production for the
privilege of being instrumentable.

Usage:
    python3 bench_range.py [path/to/librlog.so]
"""

import os
import sys
import timeit

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from rlog import RlogClient

N = 200_000
REPEAT = 5


def expensive(i):
    """Stand-in for a real argument string: formats five fields."""
    return f"m={i} n={i + 1} k={i + 2} alpha={1.0} beta={0.0}"


def build_variants(client):
    """Each variant is a callable running N iterations of one pattern.

    The loop lives inside the variant so the measurement does not include a
    per-iteration function call that none of the real call sites would pay.
    """

    def empty():
        for i in range(N):
            pass

    def unguarded():
        for i in range(N):
            client.range_push("op", "static args")
            client.range_pop()

    def guarded_literal():
        for i in range(N):
            if client.is_logging:
                client.range_push("op", "static args")
            if client.is_logging:
                client.range_pop()

    def guarded_deferred():
        for i in range(N):
            if client.is_logging:
                client.range_push("op", expensive(i))
            if client.is_logging:
                client.range_pop()

    def guarded_eager():
        for i in range(N):
            text = expensive(i)          # built before the guard is tested
            if client.is_logging:
                client.range_push("op", text)
            if client.is_logging:
                client.range_pop()

    @client.range_decorator(apiname="op", args="static args")
    def dec_literal_fn(i):
        pass

    @client.range_decorator(apiname="op", args=lambda i: expensive(i))
    def dec_deferred_fn(i):
        pass

    def dec_literal():
        for i in range(N):
            dec_literal_fn(i)

    def dec_deferred():
        for i in range(N):
            dec_deferred_fn(i)

    def ctx_eager():
        for i in range(N):
            with client.range("op", expensive(i)):
                pass

    def ctx_deferred():
        for i in range(N):
            with client.range("op", lambda: expensive(i)):
                pass

    return [
        ("empty loop", empty),
        ("raw calls, unguarded", unguarded),
        ("raw calls, guarded, literal", guarded_literal),
        ("raw calls, guarded, eager f-string", guarded_eager),
        ("raw calls, guarded, deferred", guarded_deferred),
        ("decorator, literal args", dec_literal),
        ("decorator, deferred lambda", dec_deferred),
        ("context manager, eager f-string", ctx_eager),
        ("context manager, deferred lambda", ctx_deferred),
    ]


def main():
    lib = sys.argv[1] if len(sys.argv) > 1 else "librlog.so"
    client = RlogClient(lib)

    state = "true" if client.is_logging else "false (baseline, no tool attached)"
    print(f"range: logging = {state}\n")
    print(f"  {N} iterations, best of {REPEAT}\n")

    baseline = None
    for label, fn in build_variants(client):
        best = min(timeit.repeat(fn, number=1, repeat=REPEAT))
        ns = best * 1e9 / N
        if baseline is None:
            baseline = ns
            delta = ""
        else:
            delta = f"  (+{ns - baseline:5.0f} ns over empty)"
        print(f"  {label:36s} {best * 1e3:7.1f} ms  ({ns:5.0f} ns/range){delta}")


if __name__ == "__main__":
    main()

File: benchmarks/bench_range.cpp
// Copyright (C) Advanced Micro Devices, Inc.
// SPDX-License-Identifier: MIT

// Range benchmark: measures rlog::Range with expensive arguments.
//
// Every variant below formats the same five-field string. The only difference
// is WHEN that formatting happens:
//
//   eager    — the string is built at the call site and passed to the
//              constructor. Arguments are evaluated before the constructor is
//              entered, so the guard inside it is too late to help. The cost is
//              paid on every call whether a tool is attached or not.
//
//   deferred — a lambda is passed instead. The constructor invokes it only
//              after the guard passes, so with no tool attached the formatting
//              never runs at all.
//
// Phase 1 runs with no Logger registered (the production case: instrumented
// application, no profiler attached). Phase 2 registers a no-op Logger so both
// forms are actually logging, showing that the lambda costs nothing extra once
// the work has to be done anyway.

#include <rlog/client.h>
#include <rlog/Logger.h>
#include <rlog/Range.h>
#include "Hub.h"

#include <atomic>
#include <chrono>
#include <cstdio>

static std::atomic<bool> isLogging{false};

static void onActiveChanged()
{
    isLogging.store(rlog::isActive(), std::memory_order_relaxed);
}

class NullLogger : public rlog::Logger {
public:
    void mark(const char*, const char*, const char*, const char*) override {}
    void rangePush(const char*, const char*, const char*, const char*) override {}
    void rangePop() override {}
};

static const int N = 1'000'000;

static volatile int sink = 0;

typedef std::chrono::steady_clock clk;

static double run_empty()
{
    auto t0 = clk::now();
    for (int i = 0; i < N; ++i)
        sink = i;
    auto t1 = clk::now();
    return std::chrono::duration<double, std::milli>(t1 - t0).count();
}

static double run_literal()
{
    auto t0 = clk::now();
    for (int i = 0; i < N; ++i)
        rlog::Range r(isLogging, "bench", "range", "op", "static args");
    auto t1 = clk::now();
    return std::chrono::duration<double, std::milli>(t1 - t0).count();
}

static double run_eager()
{
    auto t0 = clk::now();
    for (int i = 0; i < N; ++i)
        rlog::Range r(isLogging, "bench", "range", "op",
                      rlog::fmt("m=%d n=%d k=%d alpha=%f beta=%f",
                                i, i + 1, i + 2, 1.0, 0.0).c_str());
    auto t1 = clk::now();
    return std::chrono::duration<double, std::milli>(t1 - t0).count();
}

static double run_deferred()
{
    auto t0 = clk::now();
    for (int i = 0; i < N; ++i)
        rlog::Range r(isLogging, "bench", "range", "op",
                      [&]{ return rlog::fmt("m=%d n=%d k=%d alpha=%f beta=%f",
                                            i, i + 1, i + 2, 1.0, 0.0); });
    auto t1 = clk::now();
    return std::chrono::duration<double, std::milli>(t1 - t0).count();
}

static void report(const char *label, double ms)
{
    printf("  %-28s %8.1f ms  (%7.0f ns/range)\n", label, ms, ms * 1e6 / N);
}

int main()
{
    rlog::init();
    rlog::registerActiveCallback(onActiveChanged);

    printf("range: logging = %s\n\n",
           isLogging.load(std::memory_order_relaxed) ? "true" : "false (baseline)");

    printf("no tool attached:\n");
    report("empty loop", run_empty());
    report("Range, literal args", run_literal());
    report("Range, eager fmt args", run_eager());
    report("Range, deferred lambda", run_deferred());

    NullLogger logger;
    rlog::Hub::singleton().addLogger(logger);

    printf("\nno-op Logger attached (logging = %s):\n",
           isLogging.load(std::memory_order_relaxed) ? "true" : "false");
    report("Range, literal args", run_literal());
    report("Range, eager fmt args", run_eager());
    report("Range, deferred lambda", run_deferred());

    rlog::Hub::singleton().removeLogger(logger);
    return 0;
}
